### ISSUE TITLE
More as one map on a page

### ISSUE THREAD

**Issue by hupe13** [Raruto/leaflet-elevation#190]
### Your environment
* leaflet-elevation: 2.1.0

https://github.com/Raruto/leaflet-elevation/issues/184#issue-1180770373:

> Many of the users (including me) have multiple maps on one page, this does not work. Or the acceleration of my test data is "Infinity" (with version 1.7.6 it is correct). The developer console does not show any errors. And I don't know what could be wrong.

I tested this, but I don't know, how to solve this.  At  onAdd https://github.com/Raruto/leaflet-elevation/blob/ee93c5ae4b6668650e5950fbb69c5831cc18b320/src/control.js#L228 the _data for dist and altitude for the second graph are not available. `console.log(this);` shows that.

**Comment by Raruto** [comment]
Hi Hupe,

>  the _data for dist and altitude for the second graph are not available. console.log(this); shows that.

I suspect that the problem is hidden in here (probably the `_registerHandler` function is not called correctly in subsequent maps, for this you find yourself without `dist`, `altitude`, `time`, ... values):

https://github.com/Raruto/leaflet-elevation/blob/046825c7a20f79cb9057b7f1862ec954cde3bc57/src/control.js#L729-L749

> testing version 2.1.0 I get an error:
>
> > Uncaught (in promise) ReferenceError: map is not defined
>
> When I change in leaflet-elevation.js the lines 910, 917 and 918 from "map" to "this._map" the error is gone and I get a map with an elevation chart. It looks different than before, but it works.

https://github.com/Raruto/leaflet-elevation/commit/7ed868e66a2a4ba4aa3d812bfcb3725638eb2395 should start to handle this, also for the other problem you report I recommend you do your own testing with the following file: [leaflet-elevation_multiple-maps.html](https://github.com/Raruto/leaflet-elevation/blob/046825c7a20f79cb9057b7f1862ec954cde3bc57/examples/leaflet-elevation_multiple-maps.html). Please do this by following the instructions as mentioned in the [build guide](https://github.com/Raruto/leaflet-elevation#build-guide) section.

_Raruto_

**Comment by hupe13** [comment]
The modules load correctly. The problem is a Promise problem. I have learned that a Promise may occur only once. If it occurs a second time, it is ignored. 
You have to make sure that, on the second map, when the modules are already loaded, the functions of the modules are called. And here I fail - I do not know how.

**Comment by Raruto** [comment]
> I have learned that a Promise may occur only once. If it occurs a second time, it is ignored.

Yes, but the code in the consumer function (`.then`) is always invoked even if the promise has been already "fullfilled" (this way you can perform asynchronous operations without having to use hundreds of status flags and all the related `if-else` conditions).

---

Anyway, I think that is the problem here (basically this assignment is done only once):

https://github.com/Raruto/leaflet-elevation/blob/c76e10bb1d30b9e53af6a238cb8973dd256b7d93/src/control.js#L742-L743

as usual, you can also check as follows: 

```js
L.Control.Elevation.include({
  _loadModules: function(handlers) {
    // First map known classnames (eg. "Altitude" --> L.Control.Elevation.Altitude)
    handlers = handlers.map((h) => typeof h === 'string' && typeof Elevation[h] !== "undefined" ? Elevation[h] : h);
    // Then load optional classes and custom imports (eg. "Cadence" --> import('../src/handlers/cadence.js'))
    let modules = handlers
      .filter(h => typeof h === 'string' || h instanceof Promise)
      .map(file => file instanceof Promise ? file : this.import(this.__modulesFolder + file.toLowerCase() + '.js'));
    return Promise.all(modules).then((m) => {
      _.each(m, (exported) => {
        let fn = Object.keys(exported)[0];
        if (fn) {
          handlers[fn] = Elevation[fn] = (Elevation[fn] ?? exported[fn]);
        }
      });
      _.each(handlers, h => ["function", "object"].includes(typeof h) && this._registerHandler(h));
    });
  },
});
```

_Have a nice day,
Raruto_

**Comment by hupe13** [comment]
Ok. I put your code after this line: https://github.com/Raruto/leaflet-elevation/blob/c76e10bb1d30b9e53af6a238cb8973dd256b7d93/examples/leaflet-elevation_multiple-maps.html#L97

The first map and chart are correct. Then the developer console says: 

> Uncaught ReferenceError: Elevation is not defined
>     handlers https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:104
>     _loadModules https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:104
>     onAdd https://domain.de/raruto/leaflet-elevation/dist/leaflet-elevation.js:529
>     addTo https://unpkg.com/leaflet@1.7.1/dist/leaflet.js:5
>     addTo https://domain.de/raruto/leaflet-elevation/dist/leaflet-elevation.js:333
>     <anonymous> https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:94
> leaflet-elevation_multiple-maps.html:104:62
>     handlers https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:104
>     map self-hosted:224
>     _loadModules https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:104
>     onAdd https://domain.de/raruto/leaflet-elevation/dist/leaflet-elevation.js:529
>     addTo https://unpkg.com/leaflet@1.7.1/dist/leaflet.js:5
>     addTo https://domain.de/raruto/leaflet-elevation/dist/leaflet-elevation.js:333
>     <anonym> https://domain.de/raruto/leaflet-elevation/examples/leaflet-elevation_multiple-maps.html:94

I get the same error on tests with my WordPress plugin.

**Comment by Raruto** [comment]
Patch released in version **2.1.1**

**Comment by hupe13** [comment]
Thank you very much!
